### PR TITLE
Optimize passes

### PR DESCRIPTION
**Pull request by fabsx00**
There are currently several passes which construct potentially large diff graphs in a single thread. In this PR we optimize these passes to make use of all cores and chunk results into smaller DiffGraphs. The primary player in this change is `ParallelIteratorExecutor`.

### CODE REVIEW

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala [hunk 1]
```diff
@@ -3,44 +3,48 @@ package io.shiftleft.semanticcpg.passes.languagespecific.fuzzyc
...
 class MethodStubCreator(cpg: Cpg) extends CpgPass(cpg) {
-  import MethodStubCreator.logger
 
   private case class NameAndSignature(name: String, signature: String)
 
   // Since the method fullNames for fuzzyc are not unique, we do not have
   // a 1to1 relation and may overwrite some values. We deem this ok for now.
-  private var methodFullNameToNode = Map[String, nodes.MethodBase]()
-  private var methodToParameterCount = Map[NameAndSignature, Int]()
+  private var methodFullNameToNode = scala.collection.concurrent.TrieMap[String, nodes.MethodBase]()
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala]:
```suggestion
  private var methodFullNameToNode = TrieMap.empty[String, nodes.MethodBase]
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala [hunk 2]
```diff
@@ -3,44 +3,48 @@ package io.shiftleft.semanticcpg.passes.languagespecific.fuzzyc
...
+  private var methodToParameterCount = scala.collection.concurrent.TrieMap[NameAndSignature, Int]()
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala]:
```suggestion
  private var methodToParameterCount = TrieMap.empty[NameAndSignature, Int]
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala [hunk 3]
```diff
@@ -3,44 +3,48 @@ package io.shiftleft.semanticcpg.passes.languagespecific.fuzzyc
...
-    // will only work for C because in C, name always equals full name.
-    methodToParameterCount.foreach {
-      case (NameAndSignature(name, signature), parameterCount) =>
-        methodFullNameToNode.get(name) match {
-          case None =>
-            createMethodStub(name, name, signature, parameterCount, dstGraph)
-          case _ =>
-        }
-
+    val CHUNK_SIZE = 128
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala]:
```suggestion
    val ChunkSize = 128
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala [hunk 4]
```diff
@@ -3,44 +3,48 @@ package io.shiftleft.semanticcpg.passes.languagespecific.fuzzyc
...
+    val chunks: Iterator[TrieMap[NameAndSignature, Int]] = methodToParameterCount.grouped(CHUNK_SIZE)
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala]:
```suggestion
    val chunks: Iterator[TrieMap[NameAndSignature, Int]] = methodToParameterCount.grouped(ChunkSize)
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala [hunk 5]
```diff
@@ -14,17 +16,22 @@ import io.shiftleft.semanticcpg.language._
...
   override def run(): Iterator[DiffGraph] = {
-    val dstGraph = DiffGraph.newBuilder
 
     init()
 
-    cpg.graph.V
-      .hasLabel(NodeTypes.TYPE)
-      .sideEffectWithTraverser { traverser =>
-        val typ = traverser.get.asInstanceOf[nodes.Type]
+    val CHUNK_SIZE = 128
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala]:
```suggestion
    val ChunkSize = 128
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala [hunk 6]
```diff
@@ -14,17 +16,22 @@ import io.shiftleft.semanticcpg.language._
   */
 class TypeDeclStubCreator(cpg: Cpg) extends CpgPass(cpg) {
 
-  private var typeDeclFullNameToNode = Map[String, nodes.TypeDeclBase]()
+  private var typeDeclFullNameToNode =
+    scala.collection.concurrent.TrieMap[String, nodes.TypeDeclBase]()
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala]:
```suggestion
    TrieMap.empty[String, nodes.TypeDeclBase]
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala [hunk 7]
```diff
@@ -14,17 +16,22 @@ import io.shiftleft.semanticcpg.language._
...
-    cpg.graph.V
-      .hasLabel(NodeTypes.TYPE)
-      .sideEffectWithTraverser { traverser =>
-        val typ = traverser.get.asInstanceOf[nodes.Type]
+    val CHUNK_SIZE = 128
+    new ParallelIteratorExecutor[List[Vertex]](
+      cpg.graph.V
+        .hasLabel(NodeTypes.TYPE)
+        .l
+        .grouped(CHUNK_SIZE)).map { group =>
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala]:
```suggestion
        .grouped(ChunkSize)).map { group =>
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/methoddecorations/MethodDecoratorPass.scala [hunk 8]
```diff
@@ -23,12 +23,17 @@ class MethodDecoratorPass(cpg: Cpg) extends CpgPass(cpg) {
   private[this] var loggedDeprecatedWarning = false
   private[this] var loggedMissingTypeFullName = false
 
-  override def run() = {
-    val dstGraph = DiffGraph.newBuilder
+  override def run(): Iterator[DiffGraph] = {
 
-    cpg.graph.V
+    val CHUNK_SIZE = 128
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/methoddecorations/MethodDecoratorPass.scala]:
```suggestion
    val ChunkSize = 128
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/methoddecorations/MethodDecoratorPass.scala [hunk 9]
```diff
@@ -23,12 +23,17 @@ class MethodDecoratorPass(cpg: Cpg) extends CpgPass(cpg) {
...
+    val chunks: Iterator[List[Vertex]] = cpg.graph.V
       .hasLabel(NodeTypes.METHOD_PARAMETER_IN)
-      .sideEffect {
+      .l
+      .grouped(CHUNK_SIZE)
```
Comment (mpollmeier) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/methoddecorations/MethodDecoratorPass.scala]:
```suggestion
      .grouped(ChunkSize)
```

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala [hunk 10]
```diff
@@ -50,8 +56,8 @@ class TypeDeclStubCreator(cpg: Cpg) extends CpgPass(cpg) {
   }
 
   private def init(): Unit = {
-    cpg.typeDecl.sideEffect { typeDecl =>
+    cpg.typeDecl.l.par.foreach { typeDecl =>
```
Comment (ml86) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/TypeDeclStubCreator.scala]: Did you measure whether this improves performance? I would be very surprised if it does but that is just a feeling.

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala (hunk 1 above)
Comment (fabsx00) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/passes/languagespecific/fuzzyc/MethodStubCreator.scala]: This doesn't look right: `methodToParameterCount` disappears.

**Comment by fabsx00** [comment]
The only pass that urgently requires parallelization as well is the Linker, but that's a bigger change, so, I'll send a separate PR.

**Pull request merged**